addons/sale/models/product_template.py

```diff
@@ -62,21 +64,27 @@ def _compute_sales_count(self):
     def _check_sale_product_company(self):
         """Ensure the product is not being restricted to a single company while
         having been sold in another one in the past, as this could cause issues."""
-        target_company = self.company_id
-        if target_company:  # don't prevent writing `False`, should always work
-            subquery_products = self.env['product.product'].sudo().with_context(active_test=False)._search([('product_tmpl_id', 'in', self.ids)])
+        products_by_compagny = defaultdict(lambda: self.env['product.template'])
+        for product in self:
+            if not product.product_variant_ids:
+                # Product template with no `product_variant_ids` means it's being created.
+                continue
+            products_by_compagny[product.company_id] |= product
+        for target_company, products in products_by_compagny.items():
+            if not target_company:  # don't prevent writing `False`, should always work
+                continue
```

Comment:
could be skipped in the earlier loop with an `if not product.company_id: continue` no ?

Reply:
Indeed ! Thanks for the suggestion :)